test/gdb_script.py

```diff
@@ -0,0 +1,494 @@
+"""
+ mbed CMSIS-DAP debugger
+ Copyright (c) 2015-2015 ARM Limited
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+# Notes about this script
+# 1. This script runs inside arm-none-eabi-gdb-py
+# 2. GDB processes commands inside a queue on a worker
+#    thread.  Commands that change the state of the target
+#    should run from this queue via gdb.post_event(cb).
+# 3. After running a command that changes the state of a target
+#    in the background, like "continue&", the code needs to
+#    return so the event processing that occurs on the worker
+#    thread can continue.  No target state changes will be
+#    seen until the worker thread gets a chance to process the data.
+# 4. To make the code flow better with #3 the main test function,
+#    run_test, takes advantage of the 'yield' call.  After
+#    performing a background operation that causes the target run
+#    'yield' must be called with a timeout value.  The code
+#    will wait for the target to halt for the time specified and
+#    if no signal arrives in that time the target is explicitly
+#    halted.
+# 5. Only use packages in the standard library in this script.
+#    Since the current version of arm-none-eabi-gdb-py.exe is
+#    only 32 bit on windows, there must be a 32 bit python
+#    install for it to work.  If your primary version of python
+#    is 64 bit, you can install the 32 bit version as a non
+#    primary version.  This will allow arm-none-eabi-gdb-py.exe
+#    to work.  The only problem with this is that when pyOCD
+#    is installed through pip only the 64 bit version of
+#    pyOCD and it's dependencies will get installed.
+#    If only the standard library is used here then this
+#    script will have no external dependencies.
+
+import gdb
+from time import time
+from threading import Timer
+from functools import partial
+from random import randrange
+from itertools import product
+import traceback
+import json
+
+DEFAULT_TIMEOUT = 2.0
+STACK_OFFSET = 0x800
+TEST_RAM_OFFSET = 0x800
+MAX_TEST_SIZE = 0x1000
+MAX_BKPT = 10
+
+assert STACK_OFFSET < MAX_TEST_SIZE
+assert TEST_RAM_OFFSET < MAX_TEST_SIZE
+
+TEST_PARAM_FILE = "test_params.txt"
+TEST_RESULT_FILE = "test_results.txt"
+
+monitor_commands = [
+    "help",
+    "help reset",
+    "help halt",
+    "help resume",
+    "init",
+    "reset",
+    "reset halt",
+    "reset halt resume",
+    "halt",
+    "resume",
+    "reg",
+    "arm semihosting enable",
+    "arm semihosting disable",
+    "fawehfawoefhad"
```

Comment:
It is to make sure the GDB server doesn't crash when an invalid command is sent.  I added a comment to the code to indicate this.